File: day18_part1_2018.dart

import 'dart:io';

void main() {
  List<String> input = File('input.txt').readAsLinesSync();
  List<String> area = List.from(input);
  int minutes = 10;

  for (int i = 0; i < minutes; i++) {
    List<String> nextArea = List.from(area);

    for (int y = 0; y < area.length; y++) {
      for (int x = 0; x < area[y].length; x++) {
        int trees = 0;
        int lumberyards = 0;

        for (int dy = -1; dy <= 1; dy++) {
          for (int dx = -1; dx <= 1; dx++) {
            if (dy == 0 && dx == 0) continue;
            if (y + dy >= 0 && y + dy < area.length && x + dx >= 0 && x + dx < area[y].length) {
              if (area[y + dy][x + dx] == '|') trees++;
              if (area[y + dy][x + dx] == '#') lumberyards++;
            }
          }
        }

        if (area[y][x] == '.' && trees >= 3) {
          nextArea[y] = nextArea[y].replaceRange(x, x + 1, '|');
        } else if (area[y][x] == '|' && lumberyards >= 3) {
          nextArea[y] = nextArea[y].replaceRange(x, x + 1, '#');
        } else if (area[y][x] == '#') {
          if (trees == 0 || lumberyards == 0) {
            nextArea[y] = nextArea[y].replaceRange(x, x + 1, '.');
          }
        }
      }
    }

    area = List.from(nextArea);
  }

  int woodedAcres = 0;
  int lumberyardAcres = 0;

  for (String row in area) {
    woodedAcres += row.split('').where((element) => element == '|').length;
    lumberyardAcres += row.split('').where((element) => element == '#').length;
  }

  print(woodedAcres * lumberyardAcres);
}

File: objectivec/day18_part1_2018.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSError *error;
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        
        int rows = 50;
        int cols = 50;
        
        char area[rows][cols];
        char nextArea[rows][cols];
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                area[i][j] = [lines[i] characterAtIndex:j];
            }
        }
        
        for (int t = 0; t < 10; t++) {
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    int trees = 0;
                    int lumberyards = 0;
                    
                    for (int x = i - 1; x <= i + 1; x++) {
                        for (int y = j - 1; y <= j + 1; y++) {
                            if (x == i && y == j) {
                                continue;
                            }
                            if (x >= 0 && x < rows && y >= 0 && y < cols) {
                                if (area[x][y] == '|') {
                                    trees++;
                                } else if (area[x][y] == '#') {
                                    lumberyards++;
                                }
                            }
                        }
                    }
                    
                    if (area[i][j] == '.' && trees >= 3) {
                        nextArea[i][j] = '|';
                    } else if (area[i][j] == '|' && lumberyards >= 3) {
                        nextArea[i][j] = '#';
                    } else if (area[i][j] == '#' && (trees == 0 || lumberyards == 0)) {
                        nextArea[i][j] = '.';
                    } else {
                        nextArea[i][j] = area[i][j];
                    }
                }
            }
            
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    area[i][j] = nextArea[i][j];
                }
            }
        }
        
        int woodedAcres = 0;
        int lumberyards = 0;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (area[i][j] == '|') {
                    woodedAcres++;
                } else if (area[i][j] == '#') {
                    lumberyards++;
                }
            }
        }
        
        printf("Total resource value: %d\n", woodedAcres * lumberyards);
    }
    return 0;
}